keps/sig-instrumentation/1661-event-api-migration/README.md

```diff
@@ -0,0 +1,467 @@
+| Count int32 | |
+|  | Series EventSeries |
+| Type string | Type string |
+| } | } |
+
+Namespace in which Event will live will be equal to
+- Namespace of Regarding object, if it's namespaced,
+- NamespaceSystem, if it's not.
+
+Note that this means that if Event has both Regarding and Related objects, and only one of them is namespaced, it should be used as Regarding object.
+
+The biggest change is the semantics of the Event object in case of loops. If Series is nil it means that Event is a singleton, i.e. it happened only once and the semantics is exactly the same as currently in Events with `count = 1`. If Series is not nil it means that the Event is either beginning or the end of an Event series - equivalence of current Events with `count > 1`. Events for ongoing series have Series.State set to EventSeriesStateOngoing, while endings have Series.State set to EventSeriesStateFinished.
+
+This change is better described in the section below.
+
+
+### Performance Improvements
+
+We want to replace current behavior, where EventRecorder patches Event object every time when deduplicated Event occurs with an approach where being in the loop is treated as a state, hence Events only should be updated only when system enters or exits loop state (or is a singleton Event).
+
+Because Event object TTL in etcd we can't have above implemented cleanly, as we need to update Event objects periodically to prevent etcd garbage collection from removing ongoing series. We can use this need to update users with new data about number of occurrences.
+
+The assumption we make for deduplication logic after API changes is that Events with the same <Regarding, Action, Reason, ReportingController, ReportingInstance, Related> tuples are considered isomorphic. This allows us to define notion of "event series", which is series of isomorphic events happening not farther away from each other than some defined threshold. E.g. Events happening every second are considered a series, but Events happening every hour are not.
+
+The main goal of this change is to limit number of API requests sent to the API server to the minimum. This is important as overloading the API server can severely impact usability of the system.
+
+In the absence of errors in the system (all Pods are happily running/starting, Nodes are healthy, etc.) the number of Events is easily manageable by the system. This means that it's enough to concentrate on erroneous states and limit number of Events published when something's wrong with the cluster.
+
+There are two cases to consider: Event series, which result in ~1 API call per ~30 minutes, so won't cause a problem until there's a huge number of them; and huge number of non-series Events. To improve the latter we require that no high-cardinality data are put into any of Regarding, Action, Reason, ReportingController, ReportingInstance, Related fields. Which bound the number of Events to O(number of objects in the system^2).
```

Comment:
> To improve the latter we require that no high-cardinality data are put into any of Regarding, Action, Reason, ReportingController, ReportingInstance, Related fields. Which bound the number of Events to O(number of objects in the system^2).

other than manual inspection/review, are the helper methods/libraries being updated to help ensure this?

Reply:
I'm not aware of anything automated done in this area. That said:
- Regarding and Related are what make O(N^2) nature (no automation needed)
- ReportingController and ReportingInstance are forced by libraries
- so it's only a matter of Action and Reason if we would like to add something

I think it might be useful to have something longer-term (we've never thought about but), but I wouldn't block API GA on it.

Reply:
Thanks for catching this! I've added a small note at the end.